src/UglyToad.PdfPig/Writer/PdfDedupStreamWriter.cs

```diff
@@ -0,0 +1,200 @@
+﻿namespace UglyToad.PdfPig.Writer
+{
+using Core;
+    using Graphics.Operations;
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.IO;
+    using System.Text;
+    using Tokens;
+
+    internal class PdfDedupStreamWriter : IPdfStreamWriter
+    {
+
+        public Stream Stream { get; }
+        private int CurrentNumber { get; set; } = 1;
+        private bool DisposeStream { get; set; }
+        private const decimal DefaultVersion = 1.2m;
+        private bool Initialized { get; set; }
+        private readonly Dictionary<IndirectReference, long> offsets = new Dictionary<IndirectReference, long>();
+        private readonly Dictionary<byte[], IndirectReferenceToken> hashes = new Dictionary<byte[], IndirectReferenceToken>(new FNVByteComparison());

```

Comment:
The deduping works by comparing the byte[] data of indirect objects excluding the startobj / endobj header and footer. It does this across documents so `IToken` would be tricky as it would need to be serialized each time.

The FnvHash as key also wouldn't work as it's just a hashcode (integer) and collisions are expected. Internally the dictionary is using the FnvHash code to do the lookups (custom IEqualityComparer).  Eventually we could use a real hash of the byte[] as key but would need to determine how to handle collisions, with this setup collisions aren't possible as it does a full byte[] comparison of the data.